### ISSUE TITLE
feature request: z-ordering parameter for traces

### ISSUE THREAD

**Issue by lucgiffon** [plotly/plotly.py#2345]
Hello,

it would be very nice to have the possibilty to set the z-ordering priority of traces in plotly. I think the way matplotlib handles it is pretty handy: 

- default z-order value for a trace is 0;
- if multiple trace have the same z-order value,  matplotlib decides the ordering of traces itself (random or whatever);
- if a trace have a greater z-order than an other, it appears on top of it;
- if a trace have a smaller z-order value than the other, it appears below;

Veeeery useful. Right now I am strugling to have a marker to appear on top of a bunch of line plots and it is very annoying.

**Comment by lucgiffon** [comment]
By the way, the hack I found to overcome this problem right now is to store the trace I want to add on TOP in a globalvariable (on can think of a list if one have multiple trace to add in a certain ordering) and then, just before rendering the figure, I had the trace to the figure. Then the last added trace appears on top of all other traces.

**Comment by huidongchen** [comment]
I am encountering the same issue. I would like to add several lines on top of scatter plot generated by `px.scatter()` .I tried to reorder the traces by doing `fig.data = fig.data[::-1]` but it only changes the order shown in legend, I still have not been able to add lines on top of points after trying almost all the tricks online, which is a bit frustrating..

it would be super helpful if plotly can support something like `zorder`

**Comment by tritemio** [comment]
Another use case for this feature is related to legend ordering. Currently, if you use the order of traces to specify the legend ordering, we cannot specify a z order independently. Since legend order is either None (i.e. insertion order) or alphabetical order, it is impossible to set an arbitrary order of legend items and at the same time set and arbitrary zorder of traces.

**Comment by dmalyuta** [comment]
Has there been any movement to try to implement this?

**Comment by nicolaskruchten** [comment]
A few notes on this:

1. This would have to be implemented in Plotly.js rather than in Python
2. Traces are currently drawn in batches by type, so scatter is always drawn on top of bar, etc, and this would be quite complicated to change, so the more straightforward option would be to control the other within types first

**Comment by nicolaskruchten** [comment]
I should add that `fig.data = fig.data[::-1]` does work for me:

```python
import plotly.express as px
import plotly.graph_objects as go

fig = px.scatter(x=[1,2,3,4], y=[1,2,3,4], size=[1,2,3,4], color=["a","a","b","b"])
fig.add_trace(go.Scatter(x=[1,2,3,4], y=[1,2,3,4], mode="lines", line_color="black"))
fig.data = fig.data[::-1]
fig.show()
```

as does reordering `fig.data` like this:

```python
import plotly.express as px
import plotly.graph_objects as go

fig = px.scatter(x=[1,2,3,4], y=[1,2,3,4], size=[1,2,3,4], color=["a","a","b","b"])
fig.add_trace(go.Scatter(x=[1,2,3,4], y=[1,2,3,4], mode="lines", line_color="black"))
fig.data = (fig.data[1],fig.data[2],fig.data[0])
fig.show()
```

**Comment by JoElfner** [comment]
Any news on this? This would be a great feature!

**Comment by jaymegordo** [comment]
+ 1, would love to be able to draw scatter traces BEHIND candlesticks for example. When trying to use hover label to inspect individual candles, other lines on top prevent label from showing.

**Comment by set92** [comment]
> 
> 
> I should add that `fig.data = fig.data[::-1]` does work for me:
> 
> ```python
> import plotly.express as px
> import plotly.graph_objects as go
> 
> fig = px.scatter(x=[1,2,3,4], y=[1,2,3,4], size=[1,2,3,4], color=["a","a","b","b"])
> fig.add_trace(go.Scatter(x=[1,2,3,4], y=[1,2,3,4], mode="lines", line_color="black"))
> fig.data = fig.data[::-1]
> fig.show()
> ```
> 
> as does reordering `fig.data` like this:
> 
> ```python
> import plotly.express as px
> import plotly.graph_objects as go
> 
> fig = px.scatter(x=[1,2,3,4], y=[1,2,3,4], size=[1,2,3,4], color=["a","a","b","b"])
> fig.add_trace(go.Scatter(x=[1,2,3,4], y=[1,2,3,4], mode="lines", line_color="black"))
> fig.data = (fig.data[1],fig.data[2],fig.data[0])
> fig.show()
> ```

Seems this trick doesn't work when there is too much data. Example:
```
df = (np.random.randint(0,100,size=(10000, 2)))
color_df = np.random.randint(0,4,size=(10000, 1)).T[0]
centers = np.random.randint(0,100,size=(7, 2))

fig = px.scatter(x=df[:, 0], y=df[:, 1], color=color_df)
fig.add_scatter(
    x=centers[:, 0],
    y=centers[:, 1],
    marker=dict(size=20, color="LightSeaGreen"),
    name="Centers",
    mode="markers",
)
fig.data = (fig.data[1], fig.data[0])
fig.show()

```

I tried removing the optional parameters and use only X,Y but it keeps putting the centers below the data.

**Comment by ptim** [comment]
> Seems this trick doesn't work when there is too much data.

I'm having a similar issue with a Candlestick trace always rendering in the BG, but in my case, I see the the legend re-ordered, but the graph maintains the old `z` ordering (candles are obscured by the scatter I'm trying to put in the BG).

**Comment by benlindsay** [comment]
> I should add that fig.data = fig.data[::-1] does work for me:

This does allow you to rearrange the order of the traces, but it rearranges it both in the legend and in the plot itself. The layering of the traces still seems to be inextricably linked to the order of the legend, unless there's something I'm missing. Often, I want less important traces to be at the bottom of the legend AND at the bottom layer on the plot. I don't see a way to do this currently. +1 for z-ordering parameter suggestion :)

(BTW, I love plotly and use it all the time. Thanks for maintaining such an awesome library!)

**Comment by fleimgruber** [comment]
@set92 just a quick FYI as I got a similar problem (with "too much" data) and got it working following your lead. The hint was in https://github.com/plotly/plotly.py/issues/2345#issuecomment-671461207, i.e. the ordering is done per type. In your example in https://github.com/plotly/plotly.py/issues/2345#issuecomment-778677610 the `px.scatter` is implicitly converted to a `Scattergl` but the `add_scatter` only adds a `Scatter` object. Now when the graph object for `add_trace` has the same resulting type (`Scattergl`), this works for me:

```python
import plotly.express as px
import plotly.graph_objects as go
import numpy as np

df = np.random.randint(0, 100, size=(10000, 2))
color_df = np.random.randint(0, 4, size=(10000, 1)).T[0]
centers = np.random.randint(0, 100, size=(7, 2))

fig = px.scatter(x=df[:, 0], y=df[:, 1], color=color_df)
fig.add_trace(
    go.Scattergl(
        x=centers[:, 0],
        y=centers[:, 1],
        marker=dict(size=20, color="LightSeaGreen"),
        name="Centers",
        mode="markers",
    )
)
fig.show()
```

**Comment by nicolaskruchten** [comment]
Please note that the idea of controlling the legend order is tracked in https://github.com/plotly/plotly.py/issues/2345, and is in principle distinct from the drawing order.

**Comment by nicolaskruchten** [comment]
Update on legend ordering independently of trace ordering in `data`: this is now possible with the new `legendrank` attribute in v5.0 :) https://plotly.com/python/legend/#legend-order

**Comment by ievgennaida** [comment]
@nicolaskruchten  Thank you!
Works fine even with the subplots. 

Is it possible also to specify this for the shapes? I would like to keep sibling shapes in the same z-order.

**Comment by nicolaskruchten** [comment]
`legendrank` only impacts the legend, not the z-order, and shapes aren't linked to the legend, so this approach will not impact shapes, no.

**Comment by ievgennaida** [comment]
@nicolaskruchten 
Thank you, it's clear now, can it be also mentioned in the example documentation that it's about order, not z-order of the traces?

Basically, my problem was that when you have subplots it's hard to control order of the legends sorted.
Ex: 

- SubPlot 1 [legend 14, legend 15]
- SubPlot 2 [legend 1, legend 2, legend 3, legend 4]

Expected legend:
[legend 1, legend 2, legend 3, legend 4, legend 14, legend 15]

But in reality it was taken grouped from subplots.

This sorting problem is solved by legendrank.

But also sorting collerates with order for me:
I am sorting legend by importance and it's expected that legends on top should be rendered top most (same is applicable for the related shapes, but less critical).

Actually to achieve the top most (legend on right) I have reverted order of the traces before appending them and than reverted back by the plotly property 'revert legends order'.

Now I can append in any order to control z and rank legends to keep order.

**Comment by maegul** [comment]
Related to this, not having a z-order for a ridgeline plot caused me some trouble.  

I wanted the bottom traces to be drawn on top.  But it seems that both the stacking order for violin plots and the (implicit) z-order are both bound to the `fig.data` ordering ... it isn't possible without some unintuitive hacking (see my example result at the bottom)

See my comment on the ridgeline plots issue: https://github.com/plotly/plotly_express/issues/25#issuecomment-886375427

See my post in the community forum: https://community.plotly.com/t/ridgeline-joy-plot-without-transparency/55028?u=maegul

![Screen Shot 2021-07-26 at 2 32 46 pm](https://user-images.githubusercontent.com/13930443/126935672-8bf350f4-0a1f-4726-a451-970214288351.png)

**Comment by DJM-GM** [comment]
> 
> 
> `px.scatter` is implicitly converted to a `Scattergl` but the `add_scatter` only adds a `Scatter` object. Now when the graph object for `add_trace` has the same resulting type (`Scattergl`)

And it works for me also.

So pleased that fleimgruber found and offered the work around that, after 'wasting' a day trying to find a solution, I signed up just to say 'Thank You, fleimgruber'.

**Comment by MichauGabriel** [comment]
For the `Scattergl` trick, I also had the problem that `px.scatter` makes either `Scattergl` or `Scatter ` traces (depending on the number of points to plot?). But one can force `webgl` in all cases with 'render_mode="webgl"', so that the trick of re-ordering `fig.data` always works.

**Comment by florian6973** [comment]
Would be a very nice and useful feature

**Comment by vuk119** [comment]
Why is it still not implemented after 3 years?!

**Comment by janosh** [comment]
> Why is it still not implemented after 3 years?!

Nobody is being paid to do this. it's not happening unless you or someone else decides to donate their time.

**Comment by frozensimo** [comment]
Up! This feature would be ace to have!

**Comment by aleferna12** [comment]
Life changing feature, would love to see it implemented in the future!

**Comment by devamin** [comment]
push push

**Comment by A-Roucher** [comment]
Same here, this would be useful!

**Comment by beevabeeva** [comment]
imperative feature.

**Comment by ayjayt** [comment]
[looks like someone is working on it ](https://github.com/plotly/plotly.js/pull/6918)

**Comment by ndrezn** [comment]
Want to note, given there is so much excitement about this feature in this thread: Members of the Plotly team are currently working on this feature in `plotly.js` in https://github.com/plotly/plotly.js/pull/6918, which means it will be inherited by `plotly.py` once completed. Really excited to see this feature land!

**Comment by Farkites** [comment]
I'm happy to announce that this feature is now available from version 5.21.0 for all SVG based cartesian traces. Please see the example below:

```python
import plotly.graph_objects as go

fig = go.Figure(
    go.Bar(
        x=[0, 1, 2],
        y=[1, 2, 3]
    )

)
fig.add_trace(
    go.Scatter(
        x=[0, 1, 2],
        y=[3, 2, 1],
        zorder=-2
    )
)
fig.show()
```

![newplot (2)](https://github.com/plotly/plotly.py/assets/35932204/0e14dcc7-ec49-48d9-9c3f-d976f518b1b0)

**Comment by yotkadata** [comment]
Awesome! Thanks so much to all the developers.

**Comment by chenyulue** [comment]
> I'm happy to announce that this feature is now available from version 5.21.0 for all SVG based cartesian traces. Please see the example below:
> 
> ```python
> import plotly.graph_objects as go
> 
> fig = go.Figure(
>     go.Bar(
>         x=[0, 1, 2],
>         y=[1, 2, 3]
>     )
> 
> )
> fig.add_trace(
>     go.Scatter(
>         x=[0, 1, 2],
>         y=[3, 2, 1],
>         zorder=-2
>     )
> )
> fig.show()
> ```
> 
> ![newplot (2)](https://private-user-images.githubusercontent.com/35932204/323714222-0e14dcc7-ec49-48d9-9c3f-d976f518b1b0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQxOTUxODcsIm5iZiI6MTcxNDE5NDg4NywicGF0aCI6Ii8zNTkzMjIwNC8zMjM3MTQyMjItMGUxNGRjYzctZWM0OS00OGQ5LTljM2YtZDk3NmY1MThiMWIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI3VDA1MTQ0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY3YjNlNWMwOGQyZDAwYjJlMTZjMjk5NTU0YjljNGVlMWNkMTU4YWY1NDQzZmU2YWE5ODYxMTFjY2MwNTBjNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0ApDnrL3K-Ifc9dDnaRTdA3z8nqZ-SgxbOmlBUb8tps)

It works in web-based notebook, but not works in vs code notebook plugin. Maybe it's a problem related to vs code.

**Comment by janjoch** [comment]
Really awesome! This helps a lot.

However, it is still not supported for annotations `plotly.graph_objects.Figure.add_annotation`. Is this on the roadmap?

It would be really useful to set a watermark or "disclaimer message", for instance.

**Comment by jarrisondev** [comment]
It's still not supported for scattergl traces in plotly.js
